### DISCUSSION TITLE
Instructions for GPU Support in llama.cpp

### DISCUSSION THREAD

**Discussion by DGdev91** [oobabooga/text-generation-webui#1984] [Show and tell]
I just wanted to point out that llama.cpp has now partial GPU support for ggml processing.
There are currently 4 backends: OpenBLAS, cuBLAS (Cuda), CLBlast (OpenCL), and [an experimental fork for HipBlas (ROCm)](https://github.com/ggerganov/llama.cpp/pull/1087)

from [llama-cpp-python repo](https://github.com/abetlen/llama-cpp-python#installation-with-openblas--cublas--clblast):

### Installation with OpenBLAS / cuBLAS / CLBlast

llama.cpp supports multiple BLAS backends for faster processing. Use the FORCE_CMAKE=1 environment variable to force the use of cmake and install the pip package for the desired BLAS backend.

To install with OpenBLAS, set the LLAMA_OPENBLAS=1 environment variable before installing:
`CMAKE_ARGS="-DLLAMA_OPENBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python`

To install with cuBLAS, set the LLAMA_CUBLAS=1 environment variable before installing:
`CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python`

To install with CLBlast, set the LLAMA_CLBLAST=1 environment variable before installing:
`CMAKE_ARGS="-DLLAMA_CLBLAST=on" FORCE_CMAKE=1 pip install llama-cpp-python`

To do that, you need cmake and the required toolsets for the backend you need

**Comment by eucdee** [comment]
thats great news, my ggml models outperformed gptq by alot, now with gpu support things should get even better

**Reply by ossirytk** [reply]
The gpu seems to improve performance significantly. At least CuBlas did for me on Windows. Especially the initial model load up became a lot faster. Inference was significantly improved too. It was kind of a hassle to setup tho.

**Comment by mykeehu** [comment]
I have an i5-5300U processor without the extra GPU. I tried to enable the BLAS function with `set CMAKE_ARGS="-DLLAMA_OPENBLAS=on"` and `set CMAKE_ARGS="-DLLAMA_CLBLAST=on"`, but when I start the webui, it says `BLAS = 0`. 
Why doesn't the webui see the video card of the processor (Intel HD 5500)?

I ran this in env:
```
pip uninstall -y llama-cpp-python
set CMAKE_ARGS="-DLLAMA_OPENBLAS=on"
set FORCE_CMAKE=1
pip install llama-cpp-python --no-cache-dir
```

and then this:
```
pip uninstall -y llama-cpp-python
set CMAKE_ARGS="-DLLAMA_CLBLAST=on"
set FORCE_CMAKE=1
pip install llama-cpp-python --no-cache-dir
```

**Reply by ossirytk** [reply]
IMPORTANT. Just because these instruction were pasted in the readme and there were no errors, does **not** mean that you actually set the BLAS support on. Depending of your flavor of terminal the set command may fail quietly and you just built everything without gpu support. If setting gpu layers to ~20 does nothing, then this is probably what just happened. 

SOLUTION. Echo the env variables after setting to ensure that you actually are enabling the gpu support. 

I'm highlighting the solution here, because a bunch of people(Including me) have made this mistake and it would be nice if the solution would be highlighted here in gpu instructions.

**Reply by ossirytk** [reply]
You might want to run `set CMAKE_ARGS="-DLLAMA_OPENBLAS=on"` in your cmd or whatever and then echo the variable. If it's empty, then you have done goofed. I run this in anacoda powershell and I use
`$env:FORCE_CMAKE=1
$env:CMAKE_ARGS="-DLLAMA_CUBLAS=on"
pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir`

**Comment by ossirytk** [comment]
Note to those setting up CuBlast in windows. The Cuda toolkit installer needs 64x VisualStudio. It does not like if you have x86 Visual studio installed. It does not like if you have VisualStudio BuildTools installed. It does not like if you have multiple versions of VisualStudio installed.

So make sure that you have installed VisualStudio and uninstalled everything that might conflict. Also when installing cuda toolkit select custom installation to install all the dev tools. Make sure that Visual studio cmake is in path. Re-install llama-cpp-python package with CuBlas enabled. Make sure that the env variables are really set.

**Reply by mykeehu** [reply]
I had no problem with the CuBlas installation on the desktop, I finally managed to get it to work there. 
However, this is now a laptop without a dedicated GPU, so I wanted to install OpenBlas or CLBLAST. cmake is also installed separately with PATH. The four commands ran without error, but I don't know why webui ends up not seeing them.

**Reply by viperwasp** [reply]
Some issues I ran into setting up Cublas. Did you properly activate the right Conda environment when you did the llama.cpp part with those commands? Good luck if it's still not working I might think of some other possibilities but Ossirytk was the one who helped me.

**Comment by Pedroman1** [comment]
I have a 4090 so I am guessing cuBLAS? Sorry brand new everything so far in instructions straight forward except which blas to pick. I am guessing I could also use open blas , can I use more than one blas? Again sorry I tried to google this and search  these discussions and I don't see  clear optimal blas guide even on the link with the readme for the python repo

**Reply by viperwasp** [reply]
I will be honest. I do not know what the hell any of these really are or mean. lol I have a 4080. 
I use cuBLAS. It works well. So I assume that's your go too for Nvidia. If anyone has other info let us know.  

Also reads this post
https://github.com/oobabooga/text-generation-webui/discussions/2774
As the commands you need to run are kind of changing. No one has a really good idea of what is going on it seems.

**Reply by syaoran68** [reply]
yea i would use cuBLAS. thats what i used for my 4090

**Comment by ccbadd** [comment]
Been trying to get this to compile under Ubuntu with miniconda3 with CLBLAST for two days with no luck.  When I type env I see the variables are there and correct but always compiles with no BLAS support.  Anyone figure this out yet?

**Reply by lperezmo** [reply]
I tried compiling the python-llama-cpp with clblast on Windows and it kept doing the same. Eventually I just downloaded koboldcpp and it's working great right out of the box, maybe look into using that one. It's kind of a shame because I really like the UI of text-gen-webui, and I use it on my nvidia laptop, but it's kobold only on my amd gpu laptop.

**Reply by lperezmo** [reply]
Figured it out I think, although speed improvements are not significant

* a Reddit post I made with some options
https://reddit.com/r/LocalLLaMA/s/ITJL2HTv8s

* GitHub repo with long way around it
https://github.com/lperezmo/clblast-llama-cpp-python

**Comment by ccbadd** [comment]
I looked at the compile output and see this:"
 CMake Warning at vendor/llama.cpp/CMakeLists.txt:236 (find_package):
    By not providing "FindCLBlast.cmake" in CMAKE_MODULE_PATH this project has
    asked CMake to find a package configuration file provided by "CLBlast", but
    CMake did not find one.

    Could not find a package configuration file provided by "CLBlast" with any
    of the following names:

      CLBlastConfig.cmake
      clblast-config.cmake

    Add the installation prefix of "CLBlast" to CMAKE_PREFIX_PATH or set
    "CLBlast_DIR" to a directory containing one of the above files.  If
    "CLBlast" provides a separate development package or SDK, be sure it has
    been installed.


  CMake Warning at vendor/llama.cpp/CMakeLists.txt:246 (message):
    CLBlast not found

**Comment by St33lMouse** [comment]
Count me among the defeated. Tried to get cublas to work. Set all the variables, checked that they were set using echo, pip installed, and it just won't use the gpu.

**Reply by viperwasp** [reply]
I've got it to work on multiple machines now including first tries. Here are the exact steps I did. Not sure if all of this is completely necessary or not. 

I downloaded and installed Conda here. 
https://docs.conda.io/en/latest/miniconda.html

I than installed the Windows oobabooga-windows.zip 
I did the initial setup choosing Nvidia GPU. 

I than installed Visual Studios 2022 and you need to make sure to click the right dependence like Cmake and C++ etc. After that is done next you need to install Cuda Toolkit I installed version 12.2 yesterday on a new windows 10 machine. 

Than I ran update_windows.bat or whatever it's called in Ooba. Once that was done I opened up Anadaconda Prompt window. 
I found the correct env. 

For me it was "conda activate C:\AI\oobabooga_windows\installer_files\env" however I think you can use "conda env list" to find all environments and just use conda activate on the correct one. 

So basically you activate the right environment. And run the exact code below one line at a time. 

First Run 
pip uninstall -y llama-cpp-python

And then run each of the following one line at a time
set FORCE_CMAKE=1
set "CMAKE_ARGS=-DLLAMA_CUBLAS=on"
set "CUDAFLAGS=-arch=all -lcublas"
python -m pip install git+https://github.com/abetlen/llama-cpp-python

When that complete you should be able to boot Ooba. Choose a GGML model. And increase the slider that represents that layers you sending to the GPU or however it works. Save and reload the model. It should load with blas. I've done this process so many times now with minor issues. 

If you have various versions of visual studios and cuda installed. You may need to uninstall all of them. Reinstall each and verify that your paths are setup correctly by echoing paths like how I think you suggested. If you think you have everything installed right you simple may need to do the steps of choosing the right environment and running the lines of code I provided. 

Good luck let me know if you need any further help but I can't promise to know the answers etc.

**Reply by St33lMouse** [reply]
Alas, I'm using linux. your cudaflags variable I haven't seen before.

**Reply by viperwasp** [reply]
I have no idea how relevant anything I said is to Linux. Never touched it in my life. This may still help or it may be a completely pointless thing. Good luck I guess. lol

**Reply by ossirytk** [reply]
You might want to check the instructions [here](https://github.com/abetlen/llama-cpp-python#installation-with-openblas--cublas--clblast--metal)

**Reply by kagemartin** [reply]
Because. This. Shit. Doesn't. Work. If you've been reading papers for any amount of time you know everybody who is posting IS LYING.

**Comment by ForwardForward** [comment]
For the installation and the solution that produced the result, see user jllllllllll's post:

Problem to install llama-cpp-python on Windows 10 with GPU NVidia Support CUBlast, BLAS = 0 #721
https://github.com/abetlen/llama-cpp-python/issues/721#issuecomment-1723205068
